pkg/kv/kvserver/replica_raft_test.go

```diff
@@ -337,19 +343,29 @@ func testProposalsWithInjectedLeaseIndexAndReproposalError(t *testing.T, pipelin
 		require.NoError(t, err)
 		require.EqualValues(t, exp, n)
 	}
+
 	t.Logf("observed %d async write restarts, observed %d/%d injected aborts, %d injected illegal lease applied indexes",
 		observedAsyncWriteFailures.Load(), observedReproposalErrors.Load(), injectedReproposalErrors.Load(), insertedIllegalLeaseIndex.Load())
+	t.Logf("commands reproposed (unchanged): %d", tc.store.metrics.RaftCommandsReproposed.Count())
+	t.Logf("commands reproposed (new LAI): %d", tc.store.metrics.RaftCommandsReproposedLAI.Count())
+
 	if pipelined {
 		// If we did pipelined writes, if we needed to repropose and injected an
 		// error, this should surface as an async write failure instead.
 		require.Zero(t, observedReproposalErrors.Load())
-	}
-	if !pipelined {
+		require.Equal(t, injectedReproposalErrors.Load(), observedAsyncWriteFailures.Load())
+	} else {
 		// If we're not pipelined, we shouldn't be able to get an async write
 		// failure. This isn't testing anything about reproposals per se, rather
 		// it's validation that we're truly not doing pipelined writes.
 		require.Zero(t, observedAsyncWriteFailures.Load())
+		// All the injected reproposal errors should manifest to the transaction.
+		require.Equal(t, injectedReproposalErrors.Load(), observedReproposalErrors.Load())
 	}
+	// All incorrect lease indices should manifest either as a reproposal, or a
+	// failed reproposal (when an error is injected).
+	require.Equal(t, insertedIllegalLeaseIndex.Load(),
```

Comment:
Some of these metric comparisons might (or not) be flaky if we throw in lease transfers and retries. I've stressed the test 10k times though, and did not find any flakes. So proposing to leave as is for now, and deal with it if it flakes.

Reply:
Lease transfers shouldn't be an issue at least, `testContext` tests use a single replica afaik.